# Algoritmos - Actividad Guiada 1

Nombre: Roger Hans Rodriguez Samanez<br>
URL: https://colab.research.google.com/drive/1lJqfBFQFbfb39vkg9KlMqWj527tXl9uY?usp=sharing

http:github.com/rauxxxx/Algoritmos..


## 1. Implementación del algoritmo mediante la técnica de divide y vencerás
## Torres de Hanoi
La técnica de **divide y vencerás** se basa en resolver un problema al descomponerlo en subproblemas del mismo tipo pero de menor tamaño. Estas soluciones parciales se combinan para obtener la solución del problema original. A medida que los subproblemas se hacen más pequeños, eventualmente se llega a casos simples o triviales que pueden resolverse fácilmente. Este enfoque utiliza comúnmente la **recursividad**.

En el caso de las **Torres de Hanoi**, el juego tradicional consta de tres varillas verticales y un conjunto de discos perforados por el centro. La cantidad de discos define el nivel de dificultad. Inicialmente, todos los discos están apilados en una de las varillas, ordenados de mayor a menor tamaño (de abajo hacia arriba). Las otras dos varillas comienzan vacías. El objetivo es trasladar todos los discos desde la varilla inicial a otra vacía, cumpliendo tres reglas:

1. Solo se puede mover un disco a la vez, y para mover uno, los demás deben estar en otros postes.
2. No se puede colocar un disco más grande sobre uno más pequeño.
3. Solo se puede mover el disco que esté en la parte superior de una pila.

Aunque existen diversas formas de llegar a la solución final, todas se basan en estas reglas y aplican estrategias alineadas con la técnica de divide y vencerás.


![image.png](attachment:f6955404-44e1-4d5e-a25a-db1035bd08ec.png)![image.png](attachment:f4abec06-b28b-4e59-af42-cfe01085a6dd.png)


In [ ]:
#Torres de Hanoi - Divide y vencerás
def Torres_Hanoi(N, desde, hasta):
  #N - Cantidad de fichas
  #desde - torre inicio
  #hasta - torre fin
  if N == 1 : #condición de fin
    print("Lleva la ficha " ,desde , " hasta ", hasta )

  else: #condición recursiva
    Torres_Hanoi(N-1, desde, 6-desde-hasta )
    print("Lleva la ficha " ,desde , " hasta ", hasta )
    Torres_Hanoi(N-1, 6-desde-hasta  , hasta )

Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3


In [ ]:
# Esta ejecución sería exactamente la que muestra en la imagen superior
Torres_Hanoi(3, 1 , 3)

Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3


In [ ]:
#probando con más fichas
Torres_Hanoi(5, 1 , 3)

Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  3  hasta  1
Lleva la ficha  2  hasta  1
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  3  hasta  2
Lleva la ficha  3  hasta  1
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3


### Sucesión de Fibonacci

La **sucesión de Fibonacci** es una secuencia numérica en la que cada número es la suma de los dos anteriores. Comienza generalmente con los valores:



![image.png](attachment:9dad7083-16d7-4bfe-a236-94b335048872.png)

### Algoritmo visto en clase

In [ ]:
#Sucesión_de_Fibonacci
#Calculo del termino n-simo de la suscesión de Fibonacci
def Fibonacci(N:int):
  if N < 2:
    return 1
  else:
    return Fibonacci(N-1)+Fibonacci(N-2)

Fibonacci(8)

34

In [ ]:
import time

In [ ]:
valor = 40

In [ ]:
start = time.time()
resultado = Fibonacci(valor)
end = time.time()
metodo = "Fibonacci"
print(f"[{metodo}] f(x) =>  f({valor}) → {resultado} | Tiempo: {end - start} segundos")

[Fibonacci] f(x) =>  f(40) → 165580141 | Tiempo: 16.439416646957397 segundos


### Mejora de Algoritmo Fibonacci

In [ ]:
from functools import lru_cache

@lru_cache(maxsize=None)
def fibonacci_con_cache(n):
    if n < 2: return n
    return fibonacci(n - 1) + fibonacci(n - 2)

In [ ]:
start = time.time()
resultado = fibonacci_con_cache(valor)
end = time.time()
metodo = "Fibonacci-Cache"
print(f"[{metodo}] f(x) =>  f({valor}) → {resultado} | Tiempo: {end - start} segundos")

[Fibonacci-Cache] f(x) =>  f(40) → 102334155 | Tiempo: 0.0003216266632080078 segundos


### Conclusiones sobre la optimización del algoritmo de Fibonacci

#### Comparación de rendimiento

Al calcular el término número 40 de la sucesión de Fibonacci, se observa una diferencia drástica en el tiempo de ejecución entre las dos implementaciones:

- **Fibonacci recursivo sin optimización**
  - Tiempo: ~16.44 segundos
  - Resultado: `165580141`
  - Costoso en tiempo, debido a la gran cantidad de llamadas recursivas redundantes (`O(2^n)`).

- **Fibonacci con `@lru_cache`**
  - Tiempo: ~0.00032 segundos
  - Resultado: `102334155`
  - Altamente eficiente gracias a la **memoización**, con complejidad `O(n)`.

#### Consideraciones sobre `@lru_cache`

Si bien el decorador `@lru_cache` mejora significativamente el rendimiento, también introduce una **consideración importante sobre el uso de memoria**:

- Cada llamada única a la función queda **almacenada en memoria**, lo que puede llevar a un consumo elevado si se calculan valores muy grandes de `n` o si se invoca la función muchas veces con distintos parámetros.
- En casos extremos, el uso intensivo del cache podría provocar **errores de memoria (`MemoryError`)**.

#### Recomendaciones

- Para cálculos aislados de términos moderadamente grandes (como `n < 10⁴`), `@lru_cache` es una excelente opción.
- Para valores muy grandes o aplicaciones que requieren eficiencia en memoria, se recomienda usar una versión **iterativa** que solo almacena los últimos dos valores:

```python
def fibonacci_iterativo(n):
    if n < 2:
        return n
    a, b = 0, 1
    for _ in range(2, n + 1):
        a, b = b, a + b
    return b


## 2. Devolución de cambio por técnica voraz

La **técnica voraz** o **greedy algorithm** es una estrategia de diseño de algoritmos que construye una solución paso a paso, eligiendo en cada paso la opción que parece ser la mejor en ese momento. Esta técnica es especialmente útil en problemas de optimización, donde se busca minimizar o maximizar un valor, como el tiempo, el costo o, en este caso, el número de monedas.
![image.png](attachment:c8770eb4-f25e-468a-9087-7e622f1fc14c.png)

Uno de los ejemplos clásicos para aplicar esta técnica es el problema de la **devolución de cambio**: dado un monto a devolver y un conjunto de denominaciones de monedas disponibles, se desea determinar la menor cantidad de monedas necesarias para devolver dicho monto.

El enfoque voraz consiste en tomar la moneda de mayor valor que no exceda el monto restante, restarla del total, y repetir el proceso hasta llegar a cero. Esta solución es eficiente y funciona correctamente **cuando las denominaciones cumplen ciertas propiedades**, como ser múltiplos entre sí (por ejemplo, monedas de 1, 2, 5, 10...).

Sin embargo, **no siempre garantiza una solución óptima** si el sistema de monedas es arbitrario. Aun así, en muchos casos prácticos como el sistema decimal de monedas, este método ofrece una solución rápida y suficientemente buena.

---

### Ejemplo:
Si se desea devolver 93 soles usando monedas/billetes de 1, 2, 5, 10, 20, 50, 100, 500 el algoritmo voraz seleccionaría:

- 1 moneda de 50  
- 2 moneda de 20  
- 1 moneda de 2  
- 1 monedas de 1  

**Total: 5 monedas**


In [ ]:
#Cambio de monedas aplicando el ejemplo anterior en algoritmos
def cambio_monedas(N, SM):
  SOLUCION = [0]*len(SM)   #SOLUCION = [0,0,0,0,..]
  ValorAcumulado = 0

  for i,valor in enumerate(SM):
    monedas =  (N-ValorAcumulado)//valor
    SOLUCION[i] = monedas
    ValorAcumulado = ValorAcumulado + monedas*valor

    if ValorAcumulado == N:
      return SOLUCION
  print("No se puede calcular una solución")

In [ ]:
SM = [500,100,50,20,10,5,2,1]
cambio_monedas(93,SM)

[0, 0, 1, 2, 0, 0, 1, 1]

### Ejemplo:
Si se desea devolver 37 soles usando monedas de 1, 5, 10 y 20, el algoritmo voraz seleccionaría:

- 1 moneda de 20  
- 1 moneda de 10  
- 1 moneda de 5  
- 2 monedas de 1  

**Total: 5 monedas**

In [ ]:
SM = [20,10,5,1]
cambio_monedas(37,SM)

[1, 1, 1, 2]

## N-Reinas por técnica de vueta atrás


In [ ]:
#N Reinas - PorVuelta Atrás()

def escribe(S):
  n = len(S)
  for x in range(n):
    print("")
    for i in range(n):
      if S[i] == x+1:
        print(" X " , end="")
      else:
        print(" - ", end="")


def es_prometedora(SOLUCION,etapa):
  #print(SOLUCION)
  #Si la solución tiene dos valores iguales no es valida => Dos reinas en la misma fila
  for i in range(etapa+1):
    #print("El valor " + str(SOLUCION[i]) + " está " +  str(SOLUCION.count(SOLUCION[i])) + " veces")
    if SOLUCION.count(SOLUCION[i]) > 1:
      return False

    #Verifica las diagonales
    for j in range(i+1, etapa +1 ):
      #print("Comprobando diagonal de " + str(i) + " y " + str(j))
      if abs(i-j) == abs(SOLUCION[i]-SOLUCION[j]) : return False
  return True



def reinas(N, solucion=[], etapa=0):
  if len(solucion) == 0:
      solucion=[0 for i in range(N)]

  for i in range(1, N+1):
    solucion[etapa] = i

    if es_prometedora(solucion, etapa):
      if etapa == N-1:
        print(solucion)
        #escribe(solucion)
        print()
      else:
        reinas(N, solucion, etapa+1)
    else:
      None

    solucion[etapa] = 0

reinas(8)

[1, 5, 8, 6, 3, 7, 2, 4]

[1, 6, 8, 3, 7, 4, 2, 5]

[1, 7, 4, 6, 8, 2, 5, 3]

[1, 7, 5, 8, 2, 4, 6, 3]

[2, 4, 6, 8, 3, 1, 7, 5]

[2, 5, 7, 1, 3, 8, 6, 4]

[2, 5, 7, 4, 1, 8, 6, 3]

[2, 6, 1, 7, 4, 8, 3, 5]

[2, 6, 8, 3, 1, 4, 7, 5]

[2, 7, 3, 6, 8, 5, 1, 4]

[2, 7, 5, 8, 1, 4, 6, 3]

[2, 8, 6, 1, 3, 5, 7, 4]

[3, 1, 7, 5, 8, 2, 4, 6]

[3, 5, 2, 8, 1, 7, 4, 6]

[3, 5, 2, 8, 6, 4, 7, 1]

[3, 5, 7, 1, 4, 2, 8, 6]

[3, 5, 8, 4, 1, 7, 2, 6]

[3, 6, 2, 5, 8, 1, 7, 4]

[3, 6, 2, 7, 1, 4, 8, 5]

[3, 6, 2, 7, 5, 1, 8, 4]

[3, 6, 4, 1, 8, 5, 7, 2]

[3, 6, 4, 2, 8, 5, 7, 1]

[3, 6, 8, 1, 4, 7, 5, 2]

[3, 6, 8, 1, 5, 7, 2, 4]

[3, 6, 8, 2, 4, 1, 7, 5]

[3, 7, 2, 8, 5, 1, 4, 6]

[3, 7, 2, 8, 6, 4, 1, 5]

[3, 8, 4, 7, 1, 6, 2, 5]

[4, 1, 5, 8, 2, 7, 3, 6]

[4, 1, 5, 8, 6, 3, 7, 2]

[4, 2, 5, 8, 6, 1, 3, 7]

[4, 2, 7, 3, 6, 8, 1, 5]

[4, 2, 7, 3, 6, 8, 5, 1]

[4, 2, 7, 5, 1, 8, 6, 3]

[4, 2, 8, 5, 7, 1, 3, 6]

[4, 2, 8, 6, 1, 3, 5, 7]

[4, 6, 1, 5, 2, 8, 3, 7]

[4, 6, 8, 2, 7, 1, 3, 5]

[4, 6, 8, 3,

## Viaje por el rio. Programación dinámica

In [ ]:
TARIFAS = [
[0,5,4,3,999,999,999],
[999,0,999,2,3,999,11],
[999,999, 0,1,999,4,10],
[999,999,999, 0,5,6,9],
[999,999, 999,999,0,999,4],
[999,999, 999,999,999,0,3],
[999,999,999,999,999,999,0]
]



################################################################
def Precios(TARIFAS):
################################################################
  #Total de Nodos
  N = len(TARIFAS[0])

  #Inicialización de la tabla de precios
  PRECIOS = [ [9999]*N for i in [9999]*N]
  RUTA = [ [""]*N for i in [""]*N]

  for i in range(0,N-1):
    RUTA[i][i] = i             #Para ir de i a i se "pasa por i"
    PRECIOS[i][i] = 0          #Para ir de i a i se se paga 0
    for j in range(i+1, N):
      MIN = TARIFAS[i][j]
      RUTA[i][j] = i

      for k in range(i, j):
        if PRECIOS[i][k] + TARIFAS[k][j] < MIN:
            MIN = min(MIN, PRECIOS[i][k] + TARIFAS[k][j] )
            RUTA[i][j] = k          #Anota que para ir de i a j hay que pasar por k
        PRECIOS[i][j] = MIN

  return PRECIOS,RUTA
################################################################


PRECIOS,RUTA = Precios(TARIFAS)
#print(PRECIOS[0][6])

print("PRECIOS")
for i in range(len(TARIFAS)):
  print(PRECIOS[i])

print("\nRUTA")
for i in range(len(TARIFAS)):
  print(RUTA[i])

#Determinar la ruta con Recursividad
def calcular_ruta(RUTA, desde, hasta):
  if desde == hasta:
    #print("Ir a :" + str(desde))
    return ""
  else:
    return str(calcular_ruta( RUTA, desde, RUTA[desde][hasta])) +  \
                ',' + \
                str(RUTA[desde][hasta] \
              )

print("\nLa ruta es:")
calcular_ruta(RUTA, 0,6)

PRECIOS
[0, 5, 4, 3, 8, 8, 11]
[9999, 0, 999, 2, 3, 8, 7]
[9999, 9999, 0, 1, 6, 4, 7]
[9999, 9999, 9999, 0, 5, 6, 9]
[9999, 9999, 9999, 9999, 0, 999, 4]
[9999, 9999, 9999, 9999, 9999, 0, 3]
[9999, 9999, 9999, 9999, 9999, 9999, 9999]

RUTA
[0, 0, 0, 0, 1, 2, 5]
['', 1, 1, 1, 1, 3, 4]
['', '', 2, 2, 3, 2, 5]
['', '', '', 3, 3, 3, 3]
['', '', '', '', 4, 4, 4]
['', '', '', '', '', 5, 5]
['', '', '', '', '', '', '']

La ruta es:


',0,2,5'

# Práctica Individual


### A) Problema: Encontrar los puntos más cercanos.
Dado un conjunto de puntos se trata de encontrar los dos puntos más cercanos.

Guía de aprendizaje:
1. Suponer en 1D, una lista de números [3403, 4537, 9089, 9746, 7259,...
2. Primer intento: Fuerza bruta
3. Calcular la complejidad. ¿Se puede mejorar?
4. Segundo intento. Aplicar Divide y Vencerás
5. Calcular la complejidad. ¿Se puede mejorar?
6. Extender el algoritmo a 2D: [(1122, 6175), (135, 4076), (7296, 2741), ...,]
7. Extender el algoritmo a 3D.

### Dos puntos más cercanos

El problema del par de puntos más cercano es  consiste en obtener  la distancia más pequeña entre dos  puntos del conjunto de puntos del espacio.

#### 1. Suponer en 1D, una lista de números [3403, 4537, 9089, 9746, 7259 ...

In [ ]:
#imports
import math
import numpy as np

In [ ]:

import random
lista1D = [random.randint(0,1000) for x in range(1000)]
lista1D

[114,
 742,
 38,
 22,
 624,
 85,
 694,
 969,
 621,
 799,
 625,
 866,
 795,
 764,
 304,
 719,
 748,
 247,
 55,
 727,
 454,
 314,
 838,
 849,
 839,
 599,
 806,
 64,
 179,
 764,
 578,
 262,
 214,
 538,
 882,
 670,
 561,
 880,
 575,
 257,
 582,
 66,
 38,
 538,
 862,
 340,
 734,
 216,
 890,
 667,
 6,
 161,
 927,
 892,
 997,
 514,
 666,
 623,
 796,
 627,
 498,
 569,
 499,
 879,
 871,
 431,
 595,
 821,
 749,
 491,
 810,
 649,
 194,
 226,
 179,
 993,
 252,
 518,
 391,
 167,
 490,
 433,
 645,
 994,
 967,
 314,
 127,
 955,
 862,
 738,
 517,
 517,
 439,
 175,
 934,
 642,
 14,
 534,
 604,
 478,
 551,
 73,
 497,
 892,
 660,
 637,
 749,
 621,
 468,
 308,
 621,
 522,
 796,
 685,
 973,
 217,
 429,
 73,
 859,
 369,
 535,
 998,
 916,
 129,
 529,
 441,
 591,
 784,
 12,
 214,
 784,
 208,
 175,
 12,
 183,
 468,
 145,
 709,
 687,
 828,
 327,
 489,
 519,
 847,
 303,
 458,
 278,
 211,
 86,
 544,
 61,
 539,
 178,
 245,
 884,
 649,
 359,
 658,
 948,
 941,
 635,
 578,
 121,
 85,
 260,
 167,
 565,
 566,
 614,
 34

#### 2. Primer intento: Fuerza bruta

In [ ]:
def puntos_mas_cercanos_fuerza_bruta(lista):
    n = len(lista)
    if n < 2: return
    diff =  np.inf
    result_pair = []
    for i in range(n-1):
        for j in range(i+1, n):
            cons_diff = abs(lista[i] - lista[j])
            if cons_diff < diff:
                diff = cons_diff
                result_pair = [lista[i], lista[j]]
    return result_pair

In [ ]:
puntos_mas_cercanos_fuerza_bruta(lista1D)

[742, 742]

#### 3. Calcular la complejidad. ¿Se puede mejorar?
El algoritmo de fuerza bruta tiene complejidad de $O(n) = (n-1)(n-1) = n^2$ , debido a los dos bucles (for) que recorren la lista.

El algoritmo sí se puede optimizar utilizando la técnica de *divide y vencerás* porque es un problema similar a un problema de ordenación.

### 4. Segundo intento. Aplicar Divide y Vencerás

In [ ]:
def calcular_distancia_base_1D(vector):
    distancia_minima = np.inf
    cantidad = len(vector)
    par_mas_cercano = []

    for i in range(cantidad - 1):
        for j in range(i + 1, cantidad):
            distancia = abs(vector[i] - vector[j])
            if distancia < distancia_minima:
                distancia_minima = distancia
                par_mas_cercano = [vector[i], vector[j]]

    return distancia_minima, par_mas_cercano

def buscar_par_mas_cercano_1D(vector):
    vector.sort()
    return dividir_y_buscar_1D(vector)

def dividir_y_buscar_1D(vector):
    cantidad = len(vector)

    if cantidad <= 3:
        return calcular_distancia_base_1D(vector)

    mitad = cantidad // 2
    izquierda = vector[:mitad]
    derecha = vector[mitad:]

    dist_izq, par_izq = dividir_y_buscar_1D(izquierda)
    dist_der, par_der = dividir_y_buscar_1D(derecha)

    menor_distancia = min(dist_izq, dist_der)
    distancia_split, par_en_split = buscar_en_zona_central_1D(vector, menor_distancia)

    if par_en_split is not None:
        return distancia_split, par_en_split

    if dist_izq < dist_der:
        return dist_izq, par_izq
    else:
        return dist_der, par_der

def buscar_en_zona_central_1D(vector, d):
    cantidad = len(vector)
    centro = cantidad // 2
    punto_medio = vector[centro]

    candidatos = [p for p in vector if punto_medio - d <= p <= punto_medio + d]
    candidatos.sort()

    mejor_par = None
    cantidad_candidatos = len(candidatos)

    for i in range(cantidad_candidatos):
        for j in range(i + 1, min(i + 7, cantidad_candidatos)):
            nueva_distancia = abs(candidatos[i] - candidatos[j])
            if nueva_distancia < d:
                d = nu


In [ ]:
distancia, puntos = calcular_distancia_base_1D(list(lista1D))
print('Distancia mínima encontrada: ', distancia)
print('Puntos: ', puntos)

Distancia mínima encontrada:  0
Puntos:  [742, 742]


#### 5. Calcular la complejidad. ¿Se puede mejorar?
La complejidad del algoritmo es de $O(n) = n(log(n))^2$ usando la técnica de  *divide y vencerás*, debido a que el algoritmo divide al arreglo de puntos en dos recursivamente y a cada uno se le aplica una ordenación que se hace en $O(n) = nlog(n)$ y luego se buscan los puntos más cercanos alrededor del punto medio en tiempo $O(n) = n$.

**¿El algoritmo se puede mejorar?**
Para este caso el algoritmo ya está bastante eficiente y no se puede mejorar porque el método *sort* ya usa una combinación de dos algoritmos de ordenación muy eficientes: merge sort e insersion sort.



#### 6. Extender el algoritmo a 2D

In [ ]:

def fuerza_bruta_puntos_2D_mas_cercanos(puntos):
    distancia_minima = float('inf')
    puntos_mas_cercanos = None

    for i in range(len(puntos)):
        for j in range(i + 1, len(puntos)):
            distancia = math.dist(puntos[i], puntos[j])
            if distancia < distancia_minima:
                distancia_minima = distancia
                puntos_mas_cercanos = (puntos[i], puntos[j])

    return distancia_minima, puntos_mas_cercanos

def buscar_puntos_2D_mas_cercanos(puntos):
    puntos_ordenados = sorted(puntos, key=lambda punto: punto[0])
    return dividir_y_buscar_puntos_2D(puntos_ordenados)

def dividir_y_buscar_puntos_2D(puntos):
    cantidad = len(puntos)

    if cantidad <= 3:
        return fuerza_bruta_puntos_2D_mas_cercanos(puntos)

    mitad = cantidad // 2
    izquierda = puntos[:mitad]
    derecha = puntos[mitad:]

    dist_izquierda, puntos_izquierda = dividir_y_buscar_puntos_2D(izquierda)
    dist_derecha, puntos_derecha = dividir_y_buscar_puntos_2D(derecha)

    distancia_minima = min(dist_izquierda, dist_derecha)

    puntos_centrales = buscar_en_franja_central(puntos, distancia_minima)

    if puntos_centrales is not None:
        return distancia_minima, puntos_centrales

    if dist_izquierda < dist_derecha:
        return dist_izquierda, puntos_izquierda
    else:
        return dist_derecha, puntos_derecha

def buscar_en_franja_central(puntos, distancia_minima):
    cantidad = len(puntos)
    centro = cantidad // 2
    x_centro = puntos[centro][0]

    candidatos = [p for p in puntos if x_centro - distancia_minima <= p[0] <= x_centro + distancia_minima]
    candidatos.sort(key=lambda punto: punto[1])  # ordenar por Y

    mejores_puntos = None

    for i in range(len(candidatos)):
        for j in range(i + 1, min(i + 7, len(candidatos))):
            distancia = math.dist(candidatos[i], candidatos[j])
            if distancia < distancia_minima:
                distancia_minima = distancia
                mejores_puntos = (candidatos[i], candidatos[j])

    return mejores_puntos

In [ ]:
lista2D = [(random.randint(0,1000), random.randint(0,1000)) for x in range(1000)]

In [ ]:
points =  [(2111, 5241), (140, 3854), (6822, 3500),]
min_distance, puntos = fuerza_bruta_puntos_2D_mas_cercanos(lista2D)

print("Puntos más cercanos:", puntos)
print("Mínima distancia:", min_distance)

Puntos más cercanos: ((946, 177), (945, 177))
Mínima distancia: 1.0


In [ ]:
points =  [(2111, 5241), (140, 3854), (6822, 3500),]
min_distance, puntos = fuerza_bruta_puntos_2D_mas_cercanos(points)

print("Puntos más cercanos:", puntos)
print("Mínima distancia:", min_distance)

Puntos más cercanos: ((2111, 5241), (140, 3854))
Mínima distancia: 2410.105806806


#### 7. Extender el algoritmo a 3D



In [ ]:
#El algoritmo anterior funciona para n dimensiones
points3D = [(1, 4, 3), (1, 2, 1), (3, 5, 6), (3, 2, 4), (5, 6, 9),(5, 2, 6)]
min_distance, puntos = fuerza_bruta_puntos_2D_mas_cercanos(points3D)

print("Puntos más cercanos:", puntos)

Puntos más cercanos: ((1, 4, 3), (1, 2, 1))
